File: src/metadata/knownIds.ts
import type { MetadataWriteTarget, SchemaDefinitionId } from "../types";
import { family7GroupFromSchemaId } from "../utils/metadataWriteTarget";
import {
  schemaDefinitionIdEquals,
  schemaDefinitionIdToken,
} from "../utils/schemaDefinitionId";

const id = (
  table: string,
  tag_id: string,
  index?: number,
): SchemaDefinitionId => ({
  table,
  tag_id,
  ...(index === undefined ? {} : { index }),
});

export const KNOWN_METADATA_IDS = {
  imageDescription: id("Exif::Main", "270"),
  artist: id("Exif::Main", "315"),
  copyright: id("Exif::Main", "33432"),
  exposureTime: id("Exif::Main", "33434"),
  fNumber: id("Exif::Main", "33437"),
  focalLength: id("Exif::Main", "37386"),
  flash: id("Exif::Main", "37385"),
  dateTimeOriginal: id("Exif::Main", "36867"),
  createDate: id("Exif::Main", "36868"),
  gpsLatitudeRef: id("GPS::Main", "1"),
  gpsLatitude: id("GPS::Main", "2"),
  gpsLongitudeRef: id("GPS::Main", "3"),
  gpsLongitude: id("GPS::Main", "4"),
  gpsAltitudeRef: id("GPS::Main", "5"),
  gpsAltitude: id("GPS::Main", "6"),
  iptcObjectName: id("IPTC::ApplicationRecord", "5"),
  iptcKeywords: id("IPTC::ApplicationRecord", "25"),
  iptcDateCreated: id("IPTC::ApplicationRecord", "55"),
  iptcTimeCreated: id("IPTC::ApplicationRecord", "60"),
  iptcDigitalCreationDate: id("IPTC::ApplicationRecord", "62"),
  iptcDigitalCreationTime: id("IPTC::ApplicationRecord", "63"),
  iptcByLine: id("IPTC::ApplicationRecord", "80"),
  iptcCity: id("IPTC::ApplicationRecord", "90"),
  iptcSubLocation: id("IPTC::ApplicationRecord", "92"),
  iptcProvinceState: id("IPTC::ApplicationRecord", "95"),
  iptcCountryCode: id("IPTC::ApplicationRecord", "100"),
  iptcCountryName: id("IPTC::ApplicationRecord", "101"),
  iptcHeadline: id("IPTC::ApplicationRecord", "105"),
  iptcCopyright: id("IPTC::ApplicationRecord", "116"),
  iptcCaption: id("IPTC::ApplicationRecord", "120"),
  xmpDescription: id("XMP::dc", "description"),
  xmpTitle: id("XMP::dc", "title"),
  xmpSubject: id("XMP::dc", "subject"),
  xmpCreator: id("XMP::dc", "creator"),
  xmpRights: id("XMP::dc", "rights"),
  xmpHierarchicalSubject: id("XMP::Lightroom", "hierarchicalSubject"),
  xmpHeadline: id("XMP::photoshop", "Headline"),
  xmpCity: id("XMP::photoshop", "City"),
  xmpState: id("XMP::photoshop", "State"),
  xmpCountry: id("XMP::photoshop", "Country"),
  xmpDateCreated: id("XMP::photoshop", "DateCreated"),
  xmpLocation: id("XMP::iptcCore", "Location"),
  xmpCountryCode: id("XMP::iptcCore", "CountryCode"),
  xmpExifDateTimeOriginal: id("XMP::exif", "DateTimeOriginal"),
  xmpCreateDate: id("XMP::xmp", "CreateDate"),
  mlibAiDescription: id("UserDefined::mlib", "AIDescription"),
  mlibAiInterpretation: id("UserDefined::mlib", "AIInterpretation"),
  mlibAiObjects: id("UserDefined::mlib", "AIObjects"),
  mlibAiOcrText: id("UserDefined::mlib", "AIOcrText"),
  mlibAiTags: id("UserDefined::mlib", "AITags"),
  mlibAiModel: id("UserDefined::mlib", "AIModel"),
  mlibAiPromptVersion: id("UserDefined::mlib", "AIPromptVersion"),
  mlibAiGeneratedAt: id("UserDefined::mlib", "AIGeneratedAt"),
  offsetTime: id("Exif::Main", "36880"),
  offsetTimeOriginal: id("Exif::Main", "36881"),
  offsetTimeDigitized: id("Exif::Main", "36882"),
  subSecTimeOriginal: id("Exif::Main", "37521"),
  subSecTimeDigitized: id("Exif::Main", "37522"),
  iptcCodedCharacterSet: id("IPTC::EnvelopeRecord", "90"),
} as const;

export const GPS_IDS = {
  latitude: KNOWN_METADATA_IDS.gpsLatitude,
  latitudeRef: KNOWN_METADATA_IDS.gpsLatitudeRef,
  longitude: KNOWN_METADATA_IDS.gpsLongitude,
  longitudeRef: KNOWN_METADATA_IDS.gpsLongitudeRef,
  altitude: KNOWN_METADATA_IDS.gpsAltitude,
  altitudeRef: KNOWN_METADATA_IDS.gpsAltitudeRef,
} as const;

export function isKnownId(
  candidate: SchemaDefinitionId,
  expected: SchemaDefinitionId,
): boolean {
  return schemaDefinitionIdEquals(candidate, expected);
}

export function knownToken(value: SchemaDefinitionId): string {
  return schemaDefinitionIdToken(value);
}

const WRITE_DEFINITIONS: Array<
  readonly [SchemaDefinitionId, group1: string, tagName: string]
> = [
  [KNOWN_METADATA_IDS.imageDescription, "IFD0", "ImageDescription"],
  [KNOWN_METADATA_IDS.artist, "IFD0", "Artist"],
  [KNOWN_METADATA_IDS.copyright, "IFD0", "Copyright"],
  [KNOWN_METADATA_IDS.dateTimeOriginal, "ExifIFD", "DateTimeOriginal"],
  [KNOWN_METADATA_IDS.createDate, "ExifIFD", "CreateDate"],
  [KNOWN_METADATA_IDS.gpsLatitudeRef, "GPS", "GPSLatitudeRef"],
  [KNOWN_METADATA_IDS.gpsLatitude, "GPS", "GPSLatitude"],
  [KNOWN_METADATA_IDS.gpsLongitudeRef, "GPS", "GPSLongitudeRef"],
  [KNOWN_METADATA_IDS.gpsLongitude, "GPS", "GPSLongitude"],
  [KNOWN_METADATA_IDS.gpsAltitudeRef, "GPS", "GPSAltitudeRef"],
  [KNOWN_METADATA_IDS.gpsAltitude, "GPS", "GPSAltitude"],
  [KNOWN_METADATA_IDS.iptcObjectName, "IPTC", "ObjectName"],
  [KNOWN_METADATA_IDS.iptcKeywords, "IPTC", "Keywords"],
  [KNOWN_METADATA_IDS.iptcDateCreated, "IPTC", "DateCreated"],
  [KNOWN_METADATA_IDS.iptcTimeCreated, "IPTC", "TimeCreated"],
  [KNOWN_METADATA_IDS.iptcDigitalCreationDate, "IPTC", "DigitalCreationDate"],
  [KNOWN_METADATA_IDS.iptcDigitalCreationTime, "IPTC", "DigitalCreationTime"],
  [KNOWN_METADATA_IDS.iptcByLine, "IPTC", "By-line"],
  [KNOWN_METADATA_IDS.iptcCity, "IPTC", "City"],
  [KNOWN_METADATA_IDS.iptcSubLocation, "IPTC", "Sub-location"],
  [KNOWN_METADATA_IDS.iptcProvinceState, "IPTC", "Province-State"],
  [KNOWN_METADATA_IDS.iptcCountryCode, "IPTC", "Country-PrimaryLocationCode"],
  [KNOWN_METADATA_IDS.iptcCountryName, "IPTC", "Country-PrimaryLocationName"],
  [KNOWN_METADATA_IDS.iptcHeadline, "IPTC", "Headline"],
  [KNOWN_METADATA_IDS.iptcCopyright, "IPTC", "CopyrightNotice"],
  [KNOWN_METADATA_IDS.iptcCaption, "IPTC", "Caption-Abstract"],
  [KNOWN_METADATA_IDS.xmpDescription, "XMP-dc", "Description"],
  [KNOWN_METADATA_IDS.xmpTitle, "XMP-dc", "Title"],
  [KNOWN_METADATA_IDS.xmpSubject, "XMP-dc", "Subject"],
  [KNOWN_METADATA_IDS.xmpCreator, "XMP-dc", "Creator"],
  [KNOWN_METADATA_IDS.xmpRights, "XMP-dc", "Rights"],
  [KNOWN_METADATA_IDS.xmpHierarchicalSubject, "XMP-lr", "HierarchicalSubject"],
  [KNOWN_METADATA_IDS.xmpHeadline, "XMP-photoshop", "Headline"],
  [KNOWN_METADATA_IDS.xmpCity, "XMP-photoshop", "City"],
  [KNOWN_METADATA_IDS.xmpState, "XMP-photoshop", "State"],
  [KNOWN_METADATA_IDS.xmpCountry, "XMP-photoshop", "Country"],
  [KNOWN_METADATA_IDS.xmpDateCreated, "XMP-photoshop", "DateCreated"],
  [KNOWN_METADATA_IDS.xmpLocation, "XMP-iptcCore", "Location"],
  [KNOWN_METADATA_IDS.xmpCountryCode, "XMP-iptcCore", "CountryCode"],
  [KNOWN_METADATA_IDS.xmpCreateDate, "XMP-xmp", "CreateDate"],
  [KNOWN_METADATA_IDS.mlibAiDescription, "XMP-mlib", "AIDescription"],
  [KNOWN_METADATA_IDS.mlibAiInterpretation, "XMP-mlib", "AIInterpretation"],
  [KNOWN_METADATA_IDS.mlibAiObjects, "XMP-mlib", "AIObjects"],
  [KNOWN_METADATA_IDS.mlibAiOcrText, "XMP-mlib", "AIOcrText"],
  [KNOWN_METADATA_IDS.mlibAiTags, "XMP-mlib", "AITags"],
  [KNOWN_METADATA_IDS.mlibAiModel, "XMP-mlib", "AIModel"],
  [KNOWN_METADATA_IDS.mlibAiPromptVersion, "XMP-mlib", "AIPromptVersion"],
  [KNOWN_METADATA_IDS.mlibAiGeneratedAt, "XMP-mlib", "AIGeneratedAt"],
];

const WRITE_NAMES = new Map<string, readonly [group1: string, tagName: string]>(
  WRITE_DEFINITIONS.map(([schemaId, group1, tagName]) => [
    schemaDefinitionIdToken(schemaId),
    [group1, tagName] as const,
  ]),
);

/** Default destinations for the closed set used by generated/GPS workflows. */
export function knownMetadataWriteTarget(
  schemaId: SchemaDefinitionId,
): MetadataWriteTarget | null {
  const names = WRITE_NAMES.get(schemaDefinitionIdToken(schemaId));
  return names
    ? {
        group1: names[0],
        group7: family7GroupFromSchemaId(schemaId),
        tag_name: names[1],
      }
    : null;
}

File: src-tauri/src/known_ids.rs
//! Exact ExifTool schema identities used by specialised application behaviour.

use crate::tag_schema::SchemaDefinitionId;

fn id(table: &str, tag_id: &str) -> SchemaDefinitionId {
    SchemaDefinitionId {
        table: table.to_owned(),
        tag_id: tag_id.to_owned(),
        index: None,
    }
}

macro_rules! known_ids {
    ($($name:ident => ($table:literal, $tag_id:literal)),+ $(,)?) => {
        $(pub fn $name() -> SchemaDefinitionId { id($table, $tag_id) })+
    };
}

known_ids! {
    image_description => ("Exif::Main", "270"), artist => ("Exif::Main", "315"),
    copyright => ("Exif::Main", "33432"), date_time_original => ("Exif::Main", "36867"),
    create_date => ("Exif::Main", "36868"), offset_time => ("Exif::Main", "36880"),
    offset_time_original => ("Exif::Main", "36881"), offset_time_digitized => ("Exif::Main", "36882"),
    sub_sec_time_original => ("Exif::Main", "37521"), sub_sec_time_digitized => ("Exif::Main", "37522"),
    iptc_coded_character_set => ("IPTC::EnvelopeRecord", "90"),
    iptc_object_name => ("IPTC::ApplicationRecord", "5"), iptc_keywords => ("IPTC::ApplicationRecord", "25"),
    iptc_date_created => ("IPTC::ApplicationRecord", "55"), iptc_time_created => ("IPTC::ApplicationRecord", "60"),
    iptc_digital_creation_date => ("IPTC::ApplicationRecord", "62"), iptc_digital_creation_time => ("IPTC::ApplicationRecord", "63"),
    iptc_by_line => ("IPTC::ApplicationRecord", "80"), iptc_city => ("IPTC::ApplicationRecord", "90"),
    iptc_sub_location => ("IPTC::ApplicationRecord", "92"), iptc_province_state => ("IPTC::ApplicationRecord", "95"),
    iptc_country_code => ("IPTC::ApplicationRecord", "100"), iptc_country_name => ("IPTC::ApplicationRecord", "101"),
    iptc_headline => ("IPTC::ApplicationRecord", "105"), iptc_copyright => ("IPTC::ApplicationRecord", "116"),
    iptc_caption => ("IPTC::ApplicationRecord", "120"),
    xmp_description => ("XMP::dc", "description"), xmp_title => ("XMP::dc", "title"),
    xmp_subject => ("XMP::dc", "subject"), xmp_creator => ("XMP::dc", "creator"),
    xmp_rights => ("XMP::dc", "rights"), xmp_hierarchical_subject => ("XMP::Lightroom", "hierarchicalSubject"),
    xmp_headline => ("XMP::photoshop", "Headline"), xmp_city => ("XMP::photoshop", "City"),
    xmp_state => ("XMP::photoshop", "State"), xmp_country => ("XMP::photoshop", "Country"),
    xmp_date_created => ("XMP::photoshop", "DateCreated"), xmp_location => ("XMP::iptcCore", "Location"),
    xmp_country_code => ("XMP::iptcCore", "CountryCode"), xmp_create_date => ("XMP::xmp", "CreateDate"),
    mlib_ai_description => ("UserDefined::mlib", "AIDescription"),
    mlib_ai_interpretation => ("UserDefined::mlib", "AIInterpretation"),
    mlib_ai_objects => ("UserDefined::mlib", "AIObjects"), mlib_ai_ocr_text => ("UserDefined::mlib", "AIOcrText"),
    mlib_ai_tags => ("UserDefined::mlib", "AITags"), mlib_ai_model => ("UserDefined::mlib", "AIModel"),
    mlib_ai_prompt_version => ("UserDefined::mlib", "AIPromptVersion"),
    mlib_ai_generated_at => ("UserDefined::mlib", "AIGeneratedAt"),
}

#[cfg(test)]
pub fn test_id(label: &str) -> SchemaDefinitionId {
    match label {
        "IFD0:ImageDescription" => image_description(), "IFD0:Artist" => artist(),
        "IFD0:Copyright" => copyright(), "ExifIFD:DateTimeOriginal" => date_time_original(),
        "ExifIFD:CreateDate" => create_date(), "XMP-dc:Description" => xmp_description(),
        "XMP-dc:Title" => xmp_title(), "XMP-dc:Subject" => xmp_subject(),
        "XMP-dc:Creator" => xmp_creator(), "XMP-dc:Rights" => xmp_rights(),
        "XMP-lr:HierarchicalSubject" => xmp_hierarchical_subject(),
        "XMP-photoshop:Headline" => xmp_headline(), "XMP-photoshop:City" => xmp_city(),
        "XMP-photoshop:State" => xmp_state(), "XMP-photoshop:Country" => xmp_country(),
        "XMP-photoshop:DateCreated" => xmp_date_created(),
        "XMP-iptcCore:Location" => xmp_location(), "XMP-iptcCore:CountryCode" => xmp_country_code(),
        "XMP-xmp:CreateDate" => xmp_create_date(), "IPTC:ObjectName" => iptc_object_name(),
        "IPTC:Keywords" => iptc_keywords(), "IPTC:DateCreated" => iptc_date_created(),
        "IPTC:TimeCreated" => iptc_time_created(), "IPTC:DigitalCreationDate" => iptc_digital_creation_date(),
        "IPTC:DigitalCreationTime" => iptc_digital_creation_time(), "IPTC:By-line" => iptc_by_line(),
        "IPTC:City" => iptc_city(), "IPTC:Sub-location" => iptc_sub_location(),
        "IPTC:Province-State" => iptc_province_state(), "IPTC:Country-PrimaryLocationCode" => iptc_country_code(),
        "IPTC:Country-PrimaryLocationName" => iptc_country_name(), "IPTC:Headline" => iptc_headline(),
        "IPTC:CopyrightNotice" => iptc_copyright(), "IPTC:Caption-Abstract" => iptc_caption(),
        "XMP-mlib:AIDescription" => mlib_ai_description(), "XMP-mlib:AIInterpretation" => mlib_ai_interpretation(),
        "XMP-mlib:AIObjects" => mlib_ai_objects(), "XMP-mlib:AIOcrText" => mlib_ai_ocr_text(),
        "XMP-mlib:AITags" => mlib_ai_tags(), "XMP-mlib:AIModel" => mlib_ai_model(),
        "XMP-mlib:AIPromptVersion" => mlib_ai_prompt_version(), "XMP-mlib:AIGeneratedAt" => mlib_ai_generated_at(),
        other => panic!("no exact test schema identity registered for {other}"),
    }
}
